runtime/src/prioritization_fee.rs

```diff
@@ -0,0 +1,358 @@
+use {
+    crate::transaction_priority_details::GetTransactionPriorityDetails,
+    solana_measure::measure,
+    solana_sdk::{
+        clock::Slot,
+        pubkey::Pubkey,
+        saturating_add_assign,
+        transaction::{SanitizedTransaction, MAX_TX_ACCOUNT_LOCKS},
+    },
+    std::collections::HashMap,
+};
+
+#[derive(Debug, Default)]
+struct PrioritizationFeeMetrics {
+    // Count of writable accounts in slot; including accounts have minimum prioritization fee equal
+    // to blocker's minimum fee, the number of these irrelevant accounts is the difference between
+    // this counter and `relevant_writable_accounts_count`.
+    total_writable_accounts_count: u64,
+
+    // Count of writeable accounts with a minimum prioritization fee higher than the minimum transaction
+    // fee for this slot.
+    relevant_writable_accounts_count: u64,
+
+    // Total prioritization fees included in this slot.
+    total_prioritization_fee: u64,
+
+    // Accumulated time spent on tracking prioritization fee for each slot.
+    total_update_elapsed_us: u64,
+}
+
+impl PrioritizationFeeMetrics {
+    fn increment_total_prioritization_fee(&mut self, val: u64) {
+        saturating_add_assign!(self.total_prioritization_fee, val);
+    }
+
+    fn increment_total_update_elapsed_us(&mut self, val: u64) {
+        saturating_add_assign!(self.total_update_elapsed_us, val);
+    }
+
+    fn report(&self, slot: Slot) {
+        datapoint_info!(
+            "block_prioritization_fee",
+            ("slot", slot as i64, i64),
+            (
+                "total_writable_accounts_count",
+                self.total_writable_accounts_count as i64,
+                i64
+            ),
+            (
+                "relevant_writable_accounts_count",
+                self.relevant_writable_accounts_count as i64,
+                i64
+            ),
+            (
+                "total_prioritization_fee",
+                self.total_prioritization_fee as i64,
+                i64
+            ),
+            (
+                "total_update_elapsed_us",
+                self.total_update_elapsed_us as i64,
+                i64
+            ),
+        );
+    }
+}
+
+pub enum PrioritizationFeeError {
+    // Not able to get account locks from sanitized transaction, which is required to update block
+    // minimum fees.
+    FailGetTransactionAccountLocks,
+
+    // Not able to read priority details, including compute-unit price, from transaction.
+    // Compute-unit price is required to update block minimum fees.
+    FailGetTransactionPriorityDetails,
+
+    // Block is already finalized, trying to finalize it again is usually unexpected
+    BlockIsAlreadyFinalized,
```

Comment:
I put this comment on Tao's code already, but..  Is there a situation where this is not a huge error? This feels like an assertion to me, rather than an error return.